event_sourcery_sqlalchemy/__init__.py

```diff
@@ -43,65 +34,38 @@ class Config(BaseModel):
     gap_retry_interval: timedelta = timedelta(seconds=0.5)
 
 
-@dataclass(repr=False)
-class SQLAlchemyBackendFactory(BackendFactory):
-    _session: Session
-    _config: Config = field(default_factory=Config)
-    _serde: Serde = field(default_factory=lambda: Serde(EventRegistry()))
-    _outbox_strategy: SqlAlchemyOutboxStorageStrategy | None = None
-
-    def build(self) -> TransactionalBackend:
-        backend = TransactionalBackend()
-        backend.serde = self._serde
-        backend.in_transaction = Dispatcher(backend.serde)
-        backend.event_store = EventStore(
-            SqlAlchemyStorageStrategy(
-                self._session,
-                backend.in_transaction,
-                self._outbox_strategy,
-            ),
-            backend.serde,
+class SQLAlchemyBackend(TransactionalBackend):
+    def __init__(self) -> None:
+        super().__init__()
+        self[Session] = not_configured(
+            "Configure backend with `.configure(session, config)`",
         )
-        backend.outbox = Outbox(
-            self._outbox_strategy or NoOutboxStorageStrategy(),
-            backend.serde,
+        self[Config] = not_configured(
+            "Configure backend with `.configure(session, config)`",
         )
-        backend.subscriber = es.subscription.SubscriptionBuilder(
-            _serde=backend.serde,
-            _strategy=SqlAlchemySubscriptionStrategy(
-                self._session, self._config.gap_retry_interval
-            ),
+        self[StorageStrategy] = lambda c: SqlAlchemyStorageStrategy(
+            c[Session],
+            c[Dispatcher],
+            c.get(SqlAlchemyOutboxStorageStrategy, None),
+        ).scoped_for_tenant(c.tenant_id)
+        self[SubscriptionStrategy] = lambda c: SqlAlchemySubscriptionStrategy(
+            c[Session],
+            c[Config].gap_retry_interval,
         )
-        return backend
 
-    def with_event_registry(self, event_registry: EventRegistry) -> Self:
-        self._serde = Serde(event_registry)
+    def configure(self, session: Session, config: Config | None = None) -> Self:
```

Comment:
WDYT about allowing `configure` to be called only once? Then, someone would have to "clone" the backend to reconfigure it or instantiate a new one.

Not a must. SQLAlchemy's session AFAIR allows for reconfiguration while lagom's container does not

Reply:
I'm not sure. Now backend is able to reconfigure. So I would be consisten here. Until it's singleton you can assume you can reconfigure and expect changes.